### PR TITLE
Implemented procedures for listing users and roles

### PR DESCRIPTION
**Pull request by OliviaYtterbrink**

### CODE REVIEW

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/AuthProcedures.java [hunk 1]
```diff
@@ -86,4 +88,63 @@ public void deleteUser( @Name( "username" ) String username ) throws IllegalCred
         }
         shiroSubject.getUserManager().deleteUser( username );
     }
+
+    @PerformsDBMS
+    @Procedure( "dbms.listUsers" )
+    public Stream<StringResult> listUsers() throws IllegalCredentialsException, IOException
+    {
+        ShiroAuthSubject shiroSubject = ShiroAuthSubject.castOrFail( authSubject );
```
Comment (lutovich) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/AuthProcedures.java:120]: This and following 4 lines are repeated quite often. Maybe extract a method?

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/RoleManager.java [hunk 2]
```diff
@@ -42,4 +44,10 @@
      * @throws IOException
      */
     void removeUserFromRole( String username, String roleName ) throws IOException;
+
+    List<String> getAllRoleNames();
```
Comment (lutovich) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/RoleManager.java]: Are role names unique? Maybe return type should be a set?

File: community/security/src/main/java/org/neo4j/server/security/auth/FileUserRepository.java [hunk 3]
```diff
@@ -180,9 +180,15 @@ public int numberOfUsers()
     }
 
     @Override
-    public boolean isValidName( String name )
+    public boolean isValidUsername( String username )
     {
-        return name.matches( "^[a-zA-Z0-9_]+$" );
+        return username.matches( "^[a-zA-Z0-9_]+$" );
```
Comment (lutovich) [inline community/security/src/main/java/org/neo4j/server/security/auth/FileUserRepository.java]: This seems like a very often called utility method, maybe precompile the regex?

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/AbstractRoleRepository.java [hunk 4]
```diff
@@ -43,21 +45,22 @@
...
+        return rolesByName.get( roleName );
     }
 
     @Override
-    public Set<String> findByUsername( String username )
+    public Set<String> getRoleNamesByUsername( String username )
     {
-        return rolesByUsername.get( username );
+        Set<String> roleNames = rolesByUsername.get( username );
+        return roleNames != null ? roleNames : Collections.EMPTY_SET;
```
Comment (lutovich) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/AbstractRoleRepository.java]: Please use `Collections.emptySet()`

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/FileRoleRepository.java [hunk 5]
```diff
@@ -60,9 +60,9 @@ public void start() throws Throwable
     }
 
     @Override
-    public boolean isValidName( String name )
+    public boolean isValidRoleName( String roleName )
     {
-        return name.matches( "^[a-zA-Z0-9_]+$" );
+        return roleName.matches( "^[a-zA-Z0-9_]+$" );
```
Comment (lutovich) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/FileRoleRepository.java]: Use precompiled regex as well

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/RoleRepository.java [hunk 6]
```diff
@@ -62,7 +63,9 @@
     int numberOfRoles();
 
     /** Utility for API consumers to tell if #create() will accept a given role name */
-    boolean isValidName( String name );
+    boolean isValidRoleName( String roleName );
 
     void removeUserFromAllRoles( String username ) throws ConcurrentModificationException, IOException;
+
+    List<String> getAllRoleNames();
```
Comment (lutovich) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/RoleRepository.java]: Same question about return type. Maybe it should be `Set`?

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/AuthProcedures.java (hunk 1 above)
Comment (fickludd) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/AuthProcedures.java:120]: I agree, and we have this on a refactoring card - check the box it if you fix it now: https://trello.com/c/RDSWbUBT/125-easy-clean-up-ideas

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/RoleManager.java (hunk 2 above)
Comment (henriknyman) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/RoleManager.java]: I am not sure this is necessary. You shouldn't get duplicates unless you manually edit the role file and create them. And the result will be converted to a stream in the procedure, so the set type will not be preserved.

File: community/security/src/main/java/org/neo4j/server/security/auth/FileUserRepository.java (hunk 3 above)
Comment (chrisvest) [inline community/security/src/main/java/org/neo4j/server/security/auth/FileUserRepository.java]: And such a simple check can most likely be made faster with just character range checks, without must cost in complexity.

File: enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/RoleManager.java (hunk 2 above)
Comment (lutovich) [inline enterprise/security/src/main/java/org/neo4j/server/security/enterprise/auth/RoleManager.java]: I think it'll still clarify the interface by making type say more about the content.

**Pull request merged**